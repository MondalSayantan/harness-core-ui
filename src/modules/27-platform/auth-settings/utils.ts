/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { get } from 'lodash-es'
import type { CreateIpAllowlistConfigProps, UpdateIpAllowlistConfigProps } from '@harnessio/react-ng-manager-client'

import type { IIPAllowlistForm } from '@auth-settings/interfaces/IPAllowlistInterface'
import { SourceType } from '@auth-settings/interfaces/IPAllowlistInterface'
import { CurrencyType } from '@common/constants/SubscriptionTypes'

export function getAmountInCurrency(currency: CurrencyType, amount: number): string {
  switch (currency) {
    case CurrencyType.USD: {
      return amount.toLocaleString('en-US', {
        style: 'currency',
        currency: 'USD'
      })
    }
  }

  return ''
}

// dividing the actual amount by 100 as amount numbers are in cents
export function getDollarAmount(amount?: number, isYearly?: boolean): number {
  if (isYearly) {
    return (amount || 0) / 1200
  }
  return (amount || 0) / 100
}

export const getErrorMessage = (error: any): string =>
  get(error, 'data.error', get(error, 'data.message', error?.message))

export const buildCreateIPAllowlistPayload = (data: IIPAllowlistForm): CreateIpAllowlistConfigProps => {
  const allowSourceType = []
  if (data.allowSourceTypeUI) {
    allowSourceType.push(SourceType.UI)
  }
  if (data.allowSourceTypeAPI) {
    allowSourceType.push(SourceType.API)
  }
  return {
    body: {
      ip_allowlist_config: {
        allowed_source_type: allowSourceType,
        description: data?.description,
        enabled: false, // always false while creating
        identifier: data.identifier,
        ip_address: data.ipAddress,
        name: data.name,
        tags: data?.tags
      }
    }
  }
}

export const buildUpdateIPAllowlistPayload = (
  data: IIPAllowlistForm,
  overrideFields?: any
): UpdateIpAllowlistConfigProps => {
  const allowSourceType = []
  if (data.allowSourceTypeUI) {
    allowSourceType.push(SourceType.UI)
  }
  if (data.allowSourceTypeAPI) {
    allowSourceType.push(SourceType.API)
  }
  return {
    'ip-config-identifier': data.identifier,
    body: {
      ip_allowlist_config: {
        allowed_source_type: allowSourceType,
        description: data?.description,
        enabled: data?.enabled,
        identifier: data.identifier,
        ip_address: data.ipAddress,
        name: data.name,
        tags: data?.tags,
        ...overrideFields
      }
    }
  }
}

/**
 * Converts minutes to higher dimensions of time such as months, days and hours
 *
 * Examples:
 * formatMinutesToHigherDimensions(150);    // Output: "2 hours 30 minutes"
 * formatMinutesToHigherDimensions(4320);   // Output: "3 days"
 * formatMinutesToHigherDimensions(45000);  // Output: "1 month 1 day 6 hours"
 * @param minutes
 * @returns a string in higher dimensions
 */
export const formatMinutesToHigherDimensions = (minutes: number | undefined): string => {
  if (minutes === undefined || minutes < 0) {
    return 'Invalid input'
  }

  const MINUTES_IN_AN_HOUR = 60
  const HOURS_IN_A_DAY = 24
  const DAYS_IN_A_MONTH = 30

  const result = []

  if (minutes >= MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY * DAYS_IN_A_MONTH) {
    const months = Math.floor(minutes / (MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY * DAYS_IN_A_MONTH))
    result.push(`${months} month${months > 1 ? 's' : ''}`)
    minutes %= MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY * DAYS_IN_A_MONTH
  }

  if (minutes >= MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY) {
    const days = Math.floor(minutes / (MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY))
    result.push(`${days} day${days > 1 ? 's' : ''}`)
    minutes %= MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY
  }

  if (minutes >= MINUTES_IN_AN_HOUR) {
    const hours = Math.floor(minutes / MINUTES_IN_AN_HOUR)
    result.push(`${hours} hour${hours > 1 ? 's' : ''}`)
    minutes %= MINUTES_IN_AN_HOUR
  }

  if (minutes > 0) {
    result.push(`${minutes} minute${minutes > 1 ? 's' : ''}`)
  }

  return result.join(' ')
}
