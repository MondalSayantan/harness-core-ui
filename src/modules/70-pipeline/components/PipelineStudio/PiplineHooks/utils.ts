/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { defaultTo, flatMapDeep, identity, map, sortBy, sortedUniq } from 'lodash-es'
import { VariableMergeServiceResponse, VariableResponseMapValue } from 'services/pipeline-ng'

export interface Params {
  metadataMap: Record<string, VariableResponseMapValue>
  localStageKeys: string[]
}

export interface AllExpressionsObj {
  expressionsList: string[]
  outputExpressions: string[]
  extraExpressions: string[]
  extraOutputExpressions: string[]
}

/**
 * Traverse over stage and find out all local fqn
 */
export function traverseStageObject(
  jsonObj: Record<string, any>,
  metadataMap: Required<VariableMergeServiceResponse>['metadataMap']
): string[] {
  const keys: string[] = []
  if (jsonObj !== null && typeof jsonObj === 'object') {
    Object.entries(jsonObj).forEach(([_key, value]) => {
      keys.push(...traverseStageObject(value, metadataMap))
    })
  } else if (metadataMap[jsonObj]) {
    keys.push(jsonObj)
  }
  return keys
}

export function pickExpressionsFromMetadataMap(
  metadataMap: Record<string, VariableResponseMapValue>,
  localStageKeys: string[],
  key: 'yamlProperties' | 'yamlOutputProperties'
): string[] {
  return sortedUniq(
    sortBy(
      map(metadataMap, (item, index) =>
        index && localStageKeys.indexOf(index) > -1
          ? defaultTo(item[key]?.localName, '')
          : defaultTo(item[key]?.fqn, '')
      ).filter(p => p),
      identity
    )
  )
}

export function pickExtraExpressionsFromMetadataMap(
  metadataMap: Record<string, VariableResponseMapValue>,
  localStageKeys: string[],
  key: 'properties' | 'outputproperties'
): string[] {
  return sortedUniq(
    sortBy(
      flatMapDeep(metadataMap, (item, index) => {
        const properties = defaultTo(item.yamlExtraProperties?.[key], [])

        return properties.map((p: Record<string, string>) =>
          index && localStageKeys.indexOf(index) > -1 ? p.localName : p.fqn
        )
      }).filter(p => p),
      identity
    )
  )
}

export const getAllExpressionsFromMetadataMap = ({ metadataMap, localStageKeys }: Params): AllExpressionsObj => ({
  expressionsList: pickExpressionsFromMetadataMap(metadataMap, localStageKeys, 'yamlProperties'),
  outputExpressions: pickExpressionsFromMetadataMap(metadataMap, localStageKeys, 'yamlOutputProperties'),
  extraExpressions: pickExtraExpressionsFromMetadataMap(metadataMap, localStageKeys, 'properties'),
  extraOutputExpressions: pickExtraExpressionsFromMetadataMap(metadataMap, localStageKeys, 'outputproperties')
})
