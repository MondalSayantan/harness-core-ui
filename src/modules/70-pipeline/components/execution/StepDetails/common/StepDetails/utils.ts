/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { ResponseApprovalInstanceResponse } from 'services/pipeline-ng'

export function encodeURIWithReservedChars(uri: string): string {
  return encodeURIComponent(uri).replace(/[;,/?:@&=+$#]/g, function (c) {
    return '%' + c.charCodeAt(0).toString(16)
  })
}

export interface ExecutionMetadataType {
  approvalInstanceId: string
  mock?: {
    data?: ResponseApprovalInstanceResponse
    loading?: boolean
  }
}
