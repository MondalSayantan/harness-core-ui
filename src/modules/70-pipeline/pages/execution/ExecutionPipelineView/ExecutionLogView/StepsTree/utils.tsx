import { defaultTo } from 'lodash-es'
import { ExecutionPipelineNode } from '@pipeline/components/ExecutionStageDiagram/ExecutionPipelineModel'
import { ExecutionNode, InterruptEffectDTO } from 'services/pipeline-ng'

export function getRetryInterrupts(step: ExecutionPipelineNode<ExecutionNode>): InterruptEffectDTO[] {
  return defaultTo(step?.item?.data?.interruptHistories, defaultTo(step?.group?.data?.interruptHistories, [])).filter(
    row => row.interruptType === 'RETRY'
  )
}
