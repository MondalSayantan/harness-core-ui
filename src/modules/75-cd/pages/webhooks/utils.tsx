/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { Scope } from '@common/interfaces/SecretsInterface'

export interface AddWebhookModalData {
  name: string
  identifier: string
  connectorRef: string
  repo: string
  folderPaths: { id: string; value: string }[] | string
}

export interface NewWebhookModalProps {
  isEdit: boolean
  initialData: AddWebhookModalData
  entityScope?: Scope
  closeModal: () => void
}

export const initialWebhookModalData = {
  name: '',
  identifier: '',
  connectorRef: '',
  repo: '',
  folderPaths: []
}
