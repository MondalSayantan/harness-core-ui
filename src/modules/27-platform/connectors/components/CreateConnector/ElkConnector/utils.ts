/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { FormData } from '@platform/connectors/interfaces/ConnectorInterface'
import { AuthTypes } from '@platform/connectors/pages/connectors/utils/ConnectorHelper'
import { UseStringsReturn } from 'framework/strings'
import type { ConnectorConfigDTO } from 'services/cd-ng'
import { ElkConnectorAuthFieldsProps } from './CreateElkConnector.types'
import { ELKConnectorFields, ElkAuthType } from './CreateElkConnector.constants'

type InitializeElkConnectorArgs = {
  prevStepData?: ConnectorConfigDTO
  accountId: string
  projectIdentifier?: string
  orgIdentifier?: string
}

export function initializeElkConnector({
  prevStepData,
  accountId,
  projectIdentifier,
  orgIdentifier
}: InitializeElkConnectorArgs): FormData {
  const defaultObj = {
    url: '',
    accountId,
    username: '',
    authType: AuthTypes.USER_PASSWORD,
    password: undefined,
    projectIdentifier,
    orgIdentifier
  }

  if (!prevStepData) {
    return defaultObj
  }
  const { spec, ...prevData } = prevStepData
  return {
    ...defaultObj,
    url: prevData?.url || spec?.controllerUrl || spec?.url || '',
    password: prevData?.password || spec?.passwordRef,
    clientSecretRef: prevData?.clientSecretRef || spec?.clientSecretRef,
    authType: prevData?.authType || spec?.authType || AuthTypes.USER_PASSWORD,
    username: prevData?.username || spec?.username || '',
    clientId: prevData?.clientId || spec?.clientId,
    apiKeyId: prevData?.apiKeyId || spec?.apiKeyId,
    apiKeyRef: prevData?.apiKeyRef || spec?.apiKeyRef
  }
}

interface GetElkAuthTypeReturn {
  label: string
  value: string
}

export const getELKAuthType = (getString: UseStringsReturn['getString']): GetElkAuthTypeReturn[] => [
  {
    label: getString('usernamePassword'),
    value: AuthTypes.USER_PASSWORD
  },
  {
    label: getString('common.apikey'),
    value: AuthTypes.API_CLIENT_TOKEN
  },
  {
    label: getString('platform.connectors.bearerToken'),
    value: AuthTypes.BEARER_TOKEN
  },
  {
    label: getString('platform.connectors.elk.noAuthentication'),
    value: ElkAuthType.NONE
  }
]

export const getAuthFields = (
  getString: UseStringsReturn['getString'],
  authType?: string
): ElkConnectorAuthFieldsProps[] | null => {
  if (!authType) {
    return null
  }

  switch (authType) {
    case AuthTypes.USER_PASSWORD:
      return [
        { name: ELKConnectorFields.USERNAME, label: getString('username'), key: ELKConnectorFields.USERNAME },
        { name: ELKConnectorFields.PASSWORD, label: getString('password'), key: ELKConnectorFields.PASSWORD }
      ]

    case AuthTypes.API_CLIENT_TOKEN:
      return [
        {
          name: ELKConnectorFields.API_KEY_ID,
          label: getString('platform.connectors.elk.apiId'),
          key: ELKConnectorFields.API_KEY_ID
        },
        {
          name: ELKConnectorFields.API_KEY_REF,
          label: getString('common.apikey'),
          key: ELKConnectorFields.API_KEY_REF
        }
      ]

    case AuthTypes.BEARER_TOKEN:
      return [
        {
          name: ELKConnectorFields.API_KEY_REF,
          label: getString('platform.connectors.bearerToken'),
          key: ELKConnectorFields.API_KEY_REF
        }
      ]

    default:
      return null
  }
}
