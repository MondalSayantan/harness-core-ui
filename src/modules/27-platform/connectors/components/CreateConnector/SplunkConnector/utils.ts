/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { FormData } from '@platform/connectors/interfaces/ConnectorInterface'
import { AuthTypes } from '@platform/connectors/pages/connectors/utils/ConnectorHelper'
import type { ConnectorConfigDTO } from 'services/cd-ng'

type InitializeSplunkConnectorArgs = {
  prevStepData?: ConnectorConfigDTO
  accountId: string
  projectIdentifier?: string
  orgIdentifier?: string
}

export function initializeSplunkConnector({
  prevStepData,
  projectIdentifier,
  accountId,
  orgIdentifier
}: InitializeSplunkConnectorArgs): FormData {
  const defaultObj = {
    url: '',
    authType: AuthTypes.USER_PASSWORD,
    username: '',
    passwordRef: undefined,
    tokenRef: undefined,
    accountId: accountId,
    projectIdentifier,
    orgIdentifier
  }

  if (!prevStepData) {
    return defaultObj
  }

  const { spec, ...prevData } = prevStepData
  return {
    ...defaultObj,
    url: prevData?.url || spec?.splunkUrl || '',
    username: prevData?.username || spec?.username,
    passwordRef: prevData?.passwordRef || spec?.passwordRef,
    authType: prevData?.authType || spec?.type || defaultObj.authType,
    tokenRef: prevData?.tokenRef || spec?.tokenRef
  }
}
