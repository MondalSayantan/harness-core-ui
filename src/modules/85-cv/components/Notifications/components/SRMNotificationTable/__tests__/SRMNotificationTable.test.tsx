/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

// Inspired by src/modules/70-pipeline/components/Notifications/__tests__/NotificationTable.test.tsx

import React from 'react'
import { act, fireEvent, render, RenderResult } from '@testing-library/react'
import { findDialogContainer, TestWrapper } from '@common/utils/testUtils'
import { defaultAppStoreValues } from '@common/utils/DefaultAppStoreData'
import SRMNotificationTable from '../SRMNotificationTable'
import type { SRMNotificationTableProps } from '../SRMNotificationTable.types'
import ConfigureMonitoredServiceAlertConditions from '../../ConfigureMonitoredServiceAlertConditions/ConfigureMonitoredServiceAlertConditions'

const args: SRMNotificationTableProps = {
  data: [
    {
      notificationRule: {
        orgIdentifier: 'test_org',
        projectIdentifier: 'test_project',
        identifier: 'NotificationRule',
        name: 'NotificationRule',
        type: 'MonitoredService',
        conditions: [
          {
            type: 'ChangeImpact',
            spec: {
              changeCategories: ['Infrastructure', 'Alert'],
              threshold: 33,
              period: '60m'
            }
          },
          {
            type: 'HealthScore',
            spec: {
              threshold: 66,
              period: '120m'
            }
          },
          {
            type: 'ChangeObserved',
            spec: {
              changeCategories: ['Infrastructure', 'Alert']
            }
          },
          {
            type: 'CodeErrors',
            spec: {
              errorTrackingEventTypes: ['Exceptions', 'LogErrors'],
              errorTrackingEventStatus: ['NewEvents', 'CriticalEvents', 'ResurfacedEvents']
            }
          }
        ],
        notificationMethod: {
          type: 'Email',
          spec: {
            recipients: ['email@harness.io']
          }
        }
      },
      enabled: true,
      createdAt: 1672063927547,
      lastModifiedAt: 1672076976438
    }
  ],
  gotoPage: jest.fn(),
  onUpdate: jest.fn(),
  totalPages: 1,
  totalItems: 1,
  pageItemCount: 1,
  pageSize: 1,
  pageIndex: 0,
  notificationRulesComponent: <ConfigureMonitoredServiceAlertConditions name={'Conditions'} />,
  handleDeleteNotification: jest.fn(),
  handleCreateNotification: jest.fn(),
  handleToggleNotification: jest.fn()
}

jest.mock('uuid', () => {
  return {
    v4: jest.fn().mockReturnValueOnce(1).mockReturnValueOnce(2).mockReturnValue(3)
  }
})

describe('Notification Table test', () => {
  let container: HTMLElement
  let getByText: RenderResult['getByText']

  beforeEach(async () => {
    const renderObj = render(
      <TestWrapper
        path="/account/:accountId/projects"
        pathParams={{ accountId: 'testAcc' }}
        defaultAppStoreValues={defaultAppStoreValues}
      >
        <SRMNotificationTable {...args} />
      </TestWrapper>
    )
    container = renderObj.container
    getByText = renderObj.getByText
  })

  test('render', () => {
    expect(container).toMatchSnapshot()
  })

  test('Delete Notfication', async () => {
    const menu = container.querySelector(`[data-icon="Options"]`)
    fireEvent.click(menu!)
    const deleteMenu = getByText('delete')
    expect(deleteMenu).toBeDefined()
    await act(async () => {
      fireEvent.click(deleteMenu!)
    })
  })

  test('New Notfication', async () => {
    const addNotification = getByText('rbac.notifications.name')
    await act(async () => {
      fireEvent.click(addNotification!)
    })
    let form = findDialogContainer()
    expect(form).toBeTruthy()
    fireEvent.click(form?.querySelector('[icon="cross"]')!)
    form = findDialogContainer()
    expect(form).not.toBeTruthy()
  })
})
