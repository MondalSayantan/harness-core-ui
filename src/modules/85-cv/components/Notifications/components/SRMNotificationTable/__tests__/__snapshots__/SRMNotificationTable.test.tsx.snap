// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification Table test render 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main"
  >
    <div
      class="Layout--horizontal StyledProps--main headerActions StyledProps--flex"
    >
      <button
        aria-label="rbac.notifications.name"
        class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--withLeftIcon Button--variation Button--variation-primary"
        id="newNotificationBtn"
        type="button"
      >
        <span
          class="bp3-icon bp3-icon-plus StyledProps--main StyledProps--padding-right-xsmall"
          icon="plus"
        >
          <svg
            data-icon="plus"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              plus
            </desc>
            <path
              d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <span
          class="bp3-button-text"
        >
          rbac.notifications.name
        </span>
      </button>
    </div>
  </div>
  <div
    class="StyledProps--font StyledProps--main content StyledProps--padding-bottom-huge"
  >
    <div
      class="TableV2--table notificationTable"
    >
      <div
        class="TableV2--header"
        role="row"
      >
        <div
          class="TableV2--cell"
          colspan="1"
          role="columnheader"
          style="width: 7%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            
          </p>
        </div>
        <div
          class="TableV2--cell"
          colspan="1"
          role="columnheader"
          style="width: 18%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            ENABLEDLABEL
          </p>
        </div>
        <div
          class="TableV2--cell"
          colspan="1"
          role="columnheader"
          style="width: 23%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            CV.NOTIFICATIONS.NOTIFICATIONNAME
          </p>
        </div>
        <div
          class="TableV2--cell"
          colspan="1"
          role="columnheader"
          style="width: 24%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            CATEGORY
          </p>
        </div>
        <div
          class="TableV2--cell"
          colspan="1"
          role="columnheader"
          style="width: 26%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            RBAC.NOTIFICATIONS.NOTIFICATIONMETHOD
          </p>
        </div>
        <div
          class="TableV2--cell"
          colspan="1"
          role="columnheader"
          style="width: 2%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            
          </p>
        </div>
      </div>
      <div
        class="TableV2--body"
      >
        <div
          class="TableV2--row TableV2--card"
          role="row"
        >
          <div
            class="TableV2--cells"
          >
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 7%;"
            >
              <div
                class="Layout--horizontal StyledProps--main"
              >
                <button
                  class="bp3-button Button--button StyledProps--font StyledProps--main toggleAccordion StyledProps--color StyledProps--color-grey600 Button--with-current-color Button--iconOnly Button--withLeftIcon Button--variation Button--variation-icon"
                  style="cursor: pointer;"
                  title="Toggle Row Expanded"
                  type="button"
                >
                  <span
                    class="bp3-icon bp3-icon-chevron-right StyledProps--main"
                    icon="chevron-right"
                  >
                    <svg
                      data-icon="chevron-right"
                      height="19"
                      viewBox="0 0 16 16"
                      width="19"
                    >
                      <desc>
                        chevron-right
                      </desc>
                      <path
                        d="M10.71 7.29l-4-4a1.003 1.003 0 00-1.42 1.42L8.59 8 5.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4-4c.18-.18.29-.43.29-.71 0-.28-.11-.53-.29-.71z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 18%;"
            >
              <label
                class="bp3-control bp3-switch StyledProps--font Switch--switch StyledProps--main"
              >
                <input
                  checked=""
                  type="checkbox"
                />
                <span
                  class="bp3-control-indicator"
                />
              </label>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 23%;"
            >
              <p
                class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--color StyledProps--color-black"
                style="--text-line-clamp: 1;"
              >
                NotificationRule
              </p>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 24%;"
            >
              <p
                class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--color StyledProps--color-black"
                style="--text-line-clamp: 1;"
              >
                Service Health
              </p>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 26%;"
            >
              <div
                class="Layout--horizontal Layout--layout-spacing-small StyledProps--main"
              >
                <span
                  class="bp3-icon StyledProps--main"
                  data-icon="main-email"
                >
                  <svg
                    height="16"
                    viewBox="0 0 14 9"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 .7v7.7l4.4-4zM13.2 0H.8L7 5.1zM9.6 4.3l4.4 4V.7zM7.3 6.1c0 .1-.2.2-.3.2s-.3 0-.3-.1L5.2 4.9.7 9h12.5L8.8 4.9z"
                      fill="#4d5969"
                    />
                  </svg>
                </span>
                <p
                  class="StyledProps--font StyledProps--main"
                >
                  Email
                </p>
              </div>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 2%;"
            >
              <span
                class="bp3-popover-wrapper"
              >
                <span
                  class="bp3-popover-target"
                >
                  <button
                    class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey350 Button--with-current-color Button--iconOnly Button--without-shadow Button--withLeftIcon Button--variation Button--variation-icon"
                    type="button"
                  >
                    <span
                      class="bp3-icon StyledProps--main"
                      data-icon="Options"
                    >
                      <svg
                        fill="currentColor"
                        height="16"
                        viewBox="0 0 4 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <circle
                          cx="2"
                          cy="2"
                          r="1.5"
                        />
                        <circle
                          cx="2"
                          cy="8"
                          r="1.5"
                        />
                        <circle
                          cx="2"
                          cy="14"
                          r="1.5"
                        />
                      </svg>
                    </span>
                  </button>
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="Layout--horizontal StyledProps--main TableV2--pagination Pagination--container StyledProps--padding-top-medium StyledProps--padding-bottom-medium StyledProps--flex StyledProps--flex-align-center-center StyledProps--flex-distribution-space-between"
      >
        <span
          class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--font-variation-small-semi"
        >
          1
           of 
          1
        </span>
        <div
          class="Layout--horizontal StyledProps--main"
        >
          <button
            aria-label="Prev"
            class="bp3-button bp3-disabled Button--button StyledProps--font StyledProps--main Pagination--roundedButton Pagination--buttonLeft Button--with-current-color Button--withLeftIcon Button--size Button--small"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              class="bp3-icon bp3-icon-arrow-left StyledProps--main StyledProps--padding-right-xsmall"
              icon="arrow-left"
            >
              <svg
                data-icon="arrow-left"
                height="12"
                viewBox="0 0 16 16"
                width="12"
              >
                <desc>
                  arrow-left
                </desc>
                <path
                  d="M13.99 6.99H4.41L7.7 3.7a1.003 1.003 0 00-1.42-1.42l-5 5a1.014 1.014 0 000 1.42l5 5a1.003 1.003 0 001.42-1.42L4.41 8.99H14c.55 0 1-.45 1-1s-.46-1-1.01-1z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
            <span
              class="bp3-button-text"
            >
              Prev
            </span>
          </button>
          <button
            class="bp3-button bp3-disabled Button--button StyledProps--font StyledProps--main Pagination--roundedButton Pagination--selected Button--with-current-color"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              class="bp3-button-text"
            >
              1
            </span>
          </button>
          <button
            aria-label="Next"
            class="bp3-button bp3-disabled Button--button StyledProps--font StyledProps--main Pagination--roundedButton Pagination--buttonRight Button--with-current-color Button--withRightIcon Button--size Button--small"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              class="bp3-button-text"
            >
              Next
            </span>
            <span
              class="bp3-icon bp3-icon-arrow-right StyledProps--main"
              icon="arrow-right"
            >
              <svg
                data-icon="arrow-right"
                height="12"
                viewBox="0 0 16 16"
                width="12"
              >
                <desc>
                  arrow-right
                </desc>
                <path
                  d="M14.7 7.29l-5-5a.965.965 0 00-.71-.3 1.003 1.003 0 00-.71 1.71l3.29 3.29H1.99c-.55 0-1 .45-1 1s.45 1 1 1h9.59l-3.29 3.29a1.003 1.003 0 001.42 1.42l5-5c.18-.18.29-.43.29-.71s-.12-.52-.3-.7z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </button>
        </div>
        <span>
          Showing 1 per page
        </span>
      </div>
    </div>
  </div>
</div>
`;
